### ISSUE TITLE
request for feedback: wrangler dev

### ISSUE THREAD

**Issue by EverlastingBugstopper** [cloudflare/wrangler#1047]
`wrangler dev` is now in alpha! If you're using it, we want to hear from you. What do you like about it, what do you think should change, have you found issues with it? Comment below!

**Comment by yet-another-dev** [comment]
On change, dev enters a never-ending rebuild cycle (issue #1078)

**Comment by yet-another-dev** [comment]
Worker updates not seen in terminal until restarting dev (issue #1082)

**Comment by deini** [comment]
I just stumbled into this. Doing a `GET` request with `body` crashes.

**Comment by yet-another-dev** [comment]
On change, cloudflared tunnel breaks (issue #1107)

**Comment by maggo** [comment]
The watcher seems to be really slow, I guess it's timeout based and not on filesystem changes?

**Comment by ackerleytng** [comment]
I'd just like to add that this is a really awesome feature. It was terrible trying to develop without feedback before.

**Comment by EverlastingBugstopper** [comment]
Hey y'all - I'm back from vacation so responses here should be a bit more frequent.

> I just stumbled into this. Doing a GET request with body crashes.
@deini do you mind filing an issue with steps to reproduce? We definitely want to support every imaginable type of HTTP request and that is 100% a bug.

> The watcher seems to be really slow, I guess it's timeout based and not on filesystem changes?
@maggo my guess is that it's not really the watcher that is slow, but rather the fact that Wrangler is required to re-build and re-upload your script on every change. We have some tentative plans to address performance but for now what we probably want to do is actually explain what Wrangler is doing and when - the output is a bit cluttered right now.

> Also it seems like it's not actually picking up changes. Testing with workers-graphql-server right now
This is troubling - would you mind filing an issue with steps to reproduce?

@ackerleytng Thanks so much for the kind words, always looking for ways to make your experience better 😄

**Comment by defjosiah** [comment]
This isn't a reproduction yet, but especially early on, wanted to give you information on a crash!

Version 1.8.0 crashes with a listener bind error:
wranger.toml
```
name = "odfiles"
type = "javascript"
account_id = "REDACTED"
workers_dev = false
route = "*odfil.es/*"
zone_id = "REDACTED"
```

run cmd:
```
RUST_BACKTRACE=1 CF_API_TOKEN=REDACTED yarn wrangler dev --host odfil.es
```

response:
```
💁  JavaScript project found. Skipping unnecessary build!
thread 'main' panicked at 'error binding to 127.0.0.1:8787: error creating server listener: Address already in use (os error 48)', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.13.2/src/server/mod.rs:124:13
stack backtrace:
💁  watching "./"
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::panicking::default_hook::{{closure}}
   4: std::panicking::default_hook
   5: std::panicking::rust_panic_with_hook
   6: rust_begin_unwind
   7: std::panicking::begin_panic_fmt
   8: hyper::server::Server<hyper::server::tcp::AddrIncoming,()>::bind
   9: <std::future::GenFuture<T> as core::future::future::Future>::poll
  10: <futures_util::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll
  11: <futures_util::future::join::Join<Fut1,Fut2> as core::future::future::Future>::poll
  12: <std::future::GenFuture<T> as core::future::future::Future>::poll
  13: tokio::runtime::enter::Enter::block_on
  14: tokio::runtime::thread_pool::ThreadPool::block_on
  15: tokio::runtime::context::enter
  16: tokio::runtime::handle::Handle::enter
  17: wrangler::commands::dev::dev
  18: wrangler::run
  19: wrangler::main
  20: std::rt::lang_start::{{closure}}
  21: std::panicking::try::do_call
  22: __rust_maybe_catch_panic
  23: std::rt::lang_start_internal
  24: main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
error Command failed with exit code 101.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by EverlastingBugstopper** [comment]
Hey @defjosiah - that error message could definitely be a bit clearer but the important bit is ` Address already in use` - looks like you already have something else running (maybe another instance of `wrangler dev`) on port 8787.

I'm not sure what OS you're on - there are a few ways of finding out which process is listening on a specific port, but I think if you find and kill that process it should work. Let me know! If not and you have more details you can share - please file a separate issue 😄

**Comment by defjosiah** [comment]
@EverlastingBugstopper that definitely fixed it. I had killed the process running on that port before starting wrangler, but with further investigation, it looks like the process I killed kept restarting itself 😆

**Comment by NicholasHazen** [comment]
Is there a simple way to detect in the code, when the work is running locally in dev mode?

**Comment by EverlastingBugstopper** [comment]
Hey @NicholasHazen - not necessarily. The best way to do it is probably to set up [environments](https://developers.cloudflare.com/workers/tooling/wrangler/configuration/environments/) with environment variables.

For instance, your `wrangler.toml` might look like this

```toml
workers_dev = true
type = "javascript"
account_id = "12657839048768728910"
vars = {
  MODE = "development",
}

[env.prod]
routes = ["example.com/*"]
zone_id = "1768940583781982"
vars = {
  MODE = "production",
}
```

Then, when you ran `wrangler dev`, in your worker, `MODE === "production"` would return `false`.

When you went to publish, you could run `wrangler dev --env prod`, and then `MODE === "production"` would return `true`.

Hope this helps!

**Comment by NicholasHazen** [comment]
That seems perfectly reasonable to me, thanks for the quick feedback.

**Comment by bigelowcc** [comment]
Hi, I need to access the original host of the request to the dev server, as the request is handled based on the subdomain. Currently I can't do that wrangler dev is rewriting it to `example.com`, or whatever I pass in.

It would be great if there was a flag to not rewrite the host. This worker only provides synthetic requests from a different backend, so what the host backend is set to doesn't need to be accessible from the worker.

Alternatively if the original host was passed through as a header that could work.

Is anything like that possible at the moment?

**Comment by bigelowcc** [comment]
Separately - I'm trying to setup the untested flow of making a request to the dev worker, which snakes its way back to localhost via a trycloudflare tunnel (or a tunnel on a custom hostname).

I can't seem to get it to work. The tunnel works fine directly, but the worker reports a 502 when trying to access it.

**Comment by hobochild** [comment]
@bigelowcc I managed to do subrequests from the worker, at least `fetch('https://google.com') works. Unfortunately `localhost` is routed to some cloudflare server, I guess this is because of the tunnel?

**Comment by EverlastingBugstopper** [comment]
@bigelowcc - I'm not sure exactly what you mean, unfortunately there's currently no way to pass in a host header. But you can pass a subdomain as part of the `--host` flag, so you could make the back end `subdomain.example.com` and that should work.

As for trying to set up the localhost routing, you can check out [this template](https://github.com/GregBrimble/wrangler-cloudflared-template) by @GregBrimble which should take care of everything until we get this functionality actually introduced to Wrangler.

**Comment by aleclarson** [comment]
> Module parse failed: Unexpected token (5:16)
You may need an appropriate loader to handle this file type.

How do I add TypeScript support?

**Comment by aleclarson** [comment]
Is there a workaround for using Node.js `--inspect` to debug my worker? 

The following command doesn't work:

```sh
NODE_OPTIONS="--inspect" wrangler dev
```

The above command fails with this error message:
```
Starting inspector on 127.0.0.1:9229 failed: address already in use
Error: failed to execute `"~/.nvm/versions/node/v12.13.0/bin/node" "~/Library/Caches/.wrangler/wranglerjs-67f8bc72e4a57eef" "--output-file=/var/folders/yt/_43z9xgn67sf7jcytzz1rt5m0000gn/T/.wranglerjs_output2eiW0" "--wasm-binding=wasm" "--webpack-config=webpack.config.js"`: exited with exit code: 12
```

**Comment by EverlastingBugstopper** [comment]
Hey @aleclarson - glad you got the first question resolved. Going to hide it for now since it is not on topic (this issue is specifically for `wrangler dev` questions/feedback).

As for the second question, Cloudflare Workers are not executed within the context of a Node runtime, and Wrangler is a completely separate tool from Node written in Rust. You can use `wrangler preview` for now to open your worker and use the inspector, but you should note that the Workers Runtime does not implement the debugger so you will only be able to use the console and network tabs. 

That being said, we do have some plans to implement `wrangler dev --inspect` which should behave similarly to `node --inspect`. You can follow along with #946 to keep up with progress towards that goal (though it's important to know that there are no guarantees on time frame for this).

**Comment by aleclarson** [comment]
@EverlastingBugstopper Good to know, thanks! Do you think Cloudflare would consider maintaining an "environment polyfill" that would make a NodeJS environment behave like a Cloudflare Worker environment? That would allow me to use the NodeJS debugger. :)

**Comment by aleclarson** [comment]
When using a custom `webpack.config.js` with source maps enabled, stack traces aren't mapping back to the original source. Will `wrangler dev` have support for this soon? Is there a workaround?

I'm using inline source maps via `devtool: "inline-cheap-module-source-map"`, and I even tried using [`browser-source-map-support.js`](https://github.com/evanw/node-source-map-support/blob/61ebf232484dbe069721bbfe36c8405ab5ac3954/browser-source-map-support.js) from `source-map-support`.

**Comment by PierBover** [comment]
`wrangler dev` is a super feature! Thanks for working on this.

I've been using it since yesterday and so far the only problem is using `console.log()`. Objects get truncated and when using the good old trick of `JSON.stringify(someObject, null, '  ')` the resulting string is not parsed correctly.

Another nitpick, there is no `cf` object in the request. Makes sense obviously but could break some features that rely on it. Would be great if there was at least a mock object in its place.

**Comment by johnelliott** [comment]
This is an **enormous** improvement to my workflow. 🎉

Here's my extremely hot take on the first 10 minutes:
- It wasn't clear at first that `--host` this wouldn't hit a local origin server. The first thing I tried was a host that resolved to 127.0.0.1. After a few minutes I reasoned that this probably worked more like the in-browser dashboard and went hunting. I see this is covered in #902 and #901 
- I'm still not sure, but I assume this does not yet have 100% compatibility with things like KV, private cache API, and behaviors from the overall hosting product e.g. page rules similarly to the browser-based environment. I have yet to find the note outlining what is part of the alpha features and what to expect to need to work around.
- Having STDOUT/console in the shell is really glorious. This is enormously better than my old setup.

> Old setup: I was using the web based interface to debug top-level error handling—such as created an exception in a returned HTML rewriter promise chain—and using my own HTTP logging setup for debugging custom cache keys and origin cache control.

**Comment by EverlastingBugstopper** [comment]
@PierBover

> wrangler dev is a super feature! Thanks for working on this.

Thanks so much for saying so! Really glad we can help out 😄 

---

> I've been using it since yesterday and so far the only problem is using console.log(). Objects get truncated

unfortunately there isn't a straightforward way to show the entire object in the console.log output because the Chrome Devtools Protocol is typically interactive. If you open the inspector with `wrangler preview` you can see what i mean - logging objects will give you a brief preview and then you have to click the dropdown to get more info. Since the terminal is linear and doesn't really have a way to display that dropdown (and getting that extra info means sending another WebSocket message) we truncate it. #946 is probably going to be the best solution to this. 

---

> when using the good old trick of JSON.stringify(someObject, null, ' ') the resulting string is not parsed correctly.

Could you file a more detailed bug report on this in a separate issue? I'm not able to reproduce

---

@johnelliott 

> This is an enormous improvement to my workflow. 🎉

So glad to hear it! Thanks for the feedback, it's really appreciated 😄 

---

> It wasn't clear at first that --host this wouldn't hit a local origin server. 

Yes, we'd love to solve this with #901 but it is not a high priority for our team right now. One of our community members/resident power users @GregBrimble has set up [an automated example](https://github.com/GregBrimble/wrangler-cloudflared-template) for you that may bridge the gap for now. Like I said, eventually we'd like to integrate this directly.

---

> I'm still not sure, but I assume this does not yet have 100% compatibility with things like KV, private cache API, and behaviors from the overall hosting product e.g. page rules. I have yet to find the note outlining what is part of the alpha features and what to expect to need to work around.

KV should work, but note that right now we do not use a separate namespace for your values unless you are using Workers Sites, this should be fixed with #1032 and is a blocker on getting this feature out of alpha. Additionally we are working on making the overall experience more realistic by adding support for things like the cache api etc. by running the code directly on our edge (right now it runs in GCS). For now, all of the same restrictions that applied to `wrangler preview` apply to `wrangler dev` - though I'm not sure those are documented.

---

> Having STDOUT/console in the shell is really glorious. This is enormously better than my old setup.

Really really glad to hear this, our goal here was to get _something_ into your hands, even if it's not quite perfect yet. Incremental improvement over perfectionism 😄

**Comment by EverlastingBugstopper** [comment]
Oh! @aleclarson I missed your comments, my apologies.

> Do you think Cloudflare would consider maintaining an "environment polyfill" that would make a NodeJS environment behave like a Cloudflare Worker environment? That would allow me to use the NodeJS debugger. :)

Unfortunately not any time soon. The Cloudflare Workers Runtime is a proprietary, closed-source implementation of the Service Workers API and it's not portable. We have some ideas around how to improve the local testing experience but making it run in Node is not on the list.

---

> When using a custom webpack.config.js with source maps enabled, stack traces aren't mapping back to the original source. Will wrangler dev have support for this soon? Is there a workaround?

There are some ideas around how to get this working and we are hashing that out internally - it will likely require some changes to the runtime for this to work smoothly. No timeline on when exactly it will be supported but it is on our radar

**Comment by arunesh90** [comment]
Would love the possibility to be able to fetch from local IP ranges. My use case is that I would like to be able to use my local elasticsearch instance for developing/debugging

**Comment by EverlastingBugstopper** [comment]
Hey @arunesh90 - check out [this repo](https://github.com/GregBrimble/wrangler-cloudflared-template) and see if it solves your use case!

**Comment by arunesh90** [comment]
@EverlastingBugstopper that seems to be for putting the local worker public via a tunnel which is not what I meant.

Currently if I try to do a HTTP request to a local IP range like `10.1.0.30` the request will return a page with a embedded YouTube video. You can see a example of this at https://cloudflareworkers.com/#:http://10.1.0.30

It's understandable why this happens normally, but now that you can run a worker locally on your own machine it might make sense to open up those local ip ranges for easier development

**Comment by EverlastingBugstopper** [comment]
That repository allows you to make your localhost public and then route your preview worker requests to what is running on localhost. Since wrangler dev is connected to Cloudflare's servers, it is impossible to pass the HTTP request through to a local IP address. The only way to put something running locally behind wrangler dev is to make it accessible over the public internet. That example uses cloudflared to do so in an automated fashion.

**Comment by arunesh90** [comment]
Ah I thought that the new dev feature actually ran the worker on the local machine and thus would be able to reach internal networks that you normally wouldn't be able to reach. 
Thanks for the clarification and the great work you're doing on wrangler 😄

**Comment by faun** [comment]
I would love to see a node.js package that shims support for the HTMLRewriter API and calls out to lol-html. Currently, running automated tests with any HTML transforms is not possible. This makes testing Cloudflare worker logic locally next to impossible. Thanks!

**Comment by defjosiah** [comment]
@faun: 
I am a definite +1 for a shim for the lol-html, but if this helps, we have this mock in our testing for the HTMLRewriter, I left the comment in there because I wanted to stress that this is not the actual API at all 😄
```
const cheerio = require('cheerio');
// Simplest possible mock of the cloudflare HTMLRewriter API. The actual version
// uses the open-source library lol-html and is a streaming parser. The actual
// bindings to the cloudflare worker JS api is not published.
// We are lucky because the cheerio "element" object has the same methods as the
// cloudflare "element" object (prepend, append, etc.). Only modification to that
// object is adding the "tagName" field to the object.
class HTMLRewriterMock {
  on(element, elementTransformer) {
    return {
      async transform(response) {
        if (response.headers.get('Content-Type') !== 'text/html') {
          return response;
        }
        const text = await response.text();
        const $ = cheerio.load(text);
        const domElement = $(element);
        domElement.tagName = element;
        $(element).tagName = element;
        elementTransformer.element(domElement);
        return new Response($.root().html(), response);
      },
    };
  }
}
```

**Comment by EverlastingBugstopper** [comment]
Hey @faun - unfortunately there are no current plans to release any shims for the Cloudflare Workers runtime. We are working on making `wrangler dev` more realistic and sturdy for integration tests which may be of use.

**Comment by faun** [comment]
Thanks for the replies. I would love to see a more capable `wrangler dev` for integration tests. In the meantime, I'm going to deploy the worker to an integration test environment as a build step and then run tests in CI against the deployed worker. Unfortunately, that makes it awkward to run those tests locally without first running the deploy step.

**Comment by PierBover** [comment]
Another issue I've found is that the default Webpack config minimizes the output so debugging is not fun with uglified var names.

So I've been using a custom config for removing the minification and adding other options (aliases, etc):

```
module.exports = {
	target: "webworker",
	entry: "./index.js",
	mode: 'production',
	optimization: {
		minimize: false
	}
	// etc
}
```

The problem with this approach is that, since it's not possible to pass env vars to Webpack, I either have to manually change the parameters whenever I want to publish the worker or maintain 2 separate configs (one for dev, one for pro).

**Comment by PierBover** [comment]
I noticed the the `webpack` type does this on every change but not the `javascript` type:

> ⬇️Installing wranglerjs...
> ⬇️Installing wasm-pack...

Which slows dev considerably.

So I made a custom config with webpack so it watches changes and builds to a folder with the `wrangler.toml` file. And then I noticed that the final bundle size is much higher than when using `wrangler dev` or `wrangler publish`.

When using Webpack in production mode the my bundle is about 167kB but when using Wrangler:

> Built successfully, built project size is 51 KiB.

What is going on here?

🤔

**Comment by coreybutler** [comment]
Just my two cents.... but I really hate the dependency on webpack. I know alot of the world loves it, but I'd prefer the developer make this choice (which I do with the straight JS option and my own preprocessor). It still leaves the dev tooling bloated though.

**Comment by coreybutler** [comment]
While I'm griping, I'd also love to see the icons go away from the logs. They're not serving any purpose the text doesn't already serve. I use the CLI because it's slimmer. Otherwise, why not just build a web UI and stream the logs there?

`</griping>`

**Comment by PierBover** [comment]
> Just my two cents.... but I really hate the dependency on webpack. I know alot of the world loves it, but I'd prefer the developer make this choice (which I do with the straight JS option and my own preprocessor). It still leaves the dev tooling bloated though.

Why can't you use any bundler/preprocessor you want?

**Comment by EverlastingBugstopper** [comment]
> Another issue I've found is that the default Webpack config minimizes the output so debugging is not fun with uglified var names.

The fix for this is source maps, issue #945 (contributions welcome)

---

> I noticed the the webpack type does this on every change but not the javascript type:

> ⬇️Installing wranglerjs...
> ⬇️Installing wasm-pack...

This is a known issue with the build step, issues #745, #849, #768 (contributions welcome)

---

> While I'm griping, I'd also love to see the icons go away from the logs.

We think they're fun! If your terminal doesn't support emojis, Wrangler won't print them. 

---

> Just my two cents.... but I really hate the dependency on webpack.

Like you said, you can use the `javascript` type and it will not require webpack. We have webpack as an option so that new users can start building complex applications without worrying about if they've properly configured their build step for Workers.

---

> Why can't you use any bundler/preprocessor you want?

You can, by setting your type to `javascript` and running whatever build steps you want before running wrangler commands. Feel free to compose whatever build step you like! That being said, we'd love to have custom builds in Wrangler, issue #820

--

### A Note on This Thread

I really appreciate all of the user feedback we're gathering here, but this thread is specifically for feedback related to `wrangler dev`. If you have issues/suggestions for other parts of Wrangler, please look in the issue tracker to see if there is current discussion, and add your two cents there. Feel free to file new issues or submit PRs to address some of the things you find difficult to use! Thanks so much 😄

**Comment by PierBover** [comment]
@EverlastingBugstopper can you comment on why a production webpack bundle (not using wrangler) is bigger than the one produced by wrangler dev with type webpack with identical webpack settings?

**Comment by EverlastingBugstopper** [comment]
@PierBover - I'll probably need to see a repo to reproduce. Hard for me to know what's going on without seeing all the moving pieces. If you could file a separate issue for that, we can take a look. Thanks! 😄

**Comment by coreybutler** [comment]
@EverlastingBugstopper - the idea of "fun" emoticons is my idea of "clutter". I'd suggest a toggle setting, giving developers a choice. A terminal may support them, but that doesn't mean they should be used.

The plugin concept around building in the dev tool might be nice, but it seems like a bit of a "recreate the wheel" concept, given all the changing choices in the JS community alone. 

In looking at the source more carefully, it looks like webpack may not be installed by default? 

Really, my biggest problem with it is the extensive node module bloat, which is the default setting for JS users. As the creator of one of the popular Node version managers, I am constantly dealing with scenarios of bloat. It's not that I'm opposed to webpack (though I'm more of a rollup person), but perhaps make it an option instead of the default. Perhaps this is more of a general wrangler issue than the dev subcommand?

**Comment by EverlastingBugstopper** [comment]
These are likely valid points, however none of these concerns have anything to do with `wrangler dev` and I would appreciate if we could move discussion to relevant issues. Thanks

**Comment by PierBover** [comment]
I'm getting this Rust error whenever I sleep my computer during `dev` or simply after a couple of minutes of not using the worker:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Protocol("Connection reset without closing handshake")', src/commands/dev/socket.rs:36:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I've been having these ever since I started using `dev`.

Is this something that's being worked on @EverlastingBugstopper ? Should I file a bug?

**Comment by grakic** [comment]
> @faun:
> I am a definite +1 for a shim for the lol-html, but if this helps, we have this mock in our testing for the HTMLRewriter ...

Thanks. Adding a bit more complete version https://github.com/dollarshaveclub/cloudworker/issues/138#issuecomment-623186542

**Comment by EverlastingBugstopper** [comment]
@PierBover - I've noticed that too and will need to design something around it, could you file a ticket? Thanks a bunch.

@grakic - you should not be getting any timeouts with `wrangler dev` - the limits should be much less strict actually. If you could file an issue with steps to reproduce (a repo link is very helpful for these) that would be great!

**Comment by tibotiber** [comment]
@EverlastingBugstopper great work on `wrangler dev` 🚀. Is there any plan to support offline dev at any point in time in the future? That would be a killer!

**Comment by tibotiber** [comment]
Added request for CORS option in #1262

**Comment by EverlastingBugstopper** [comment]
> great work on `wrangler dev`

thanks for the kind words :)

---

> Is there any plan to support offline dev at any point in time in the future?

I wish I could tell you yes, but there are no plans to make a completely local development environment for Workers.

**Comment by GregBrimble** [comment]
Hi @EverlastingBugstopper ,

As you know, I've been using `wrangler dev` pretty exhaustively since its alpha release: it's a killer feature, and I'm relying on it daily. Thank you to you and the team for your work on it.

A couple of pieces of feedback:

* I know it's not high priority, and I've worked around it with a cloudflared tunnel, but as mentioned a bit above, https://github.com/cloudflare/wrangler/issues/901 would be a nice feature for developer experience, so no one has to rely on my hacky script.

* Confirming PierBover's report of thread panics on a broken connection: https://github.com/cloudflare/wrangler/issues/1241.

* I occassionally get a couple other types of panics, but they're usually my fault with a broken worker script or something. For example, if I import a non-existent file from within a worker script, the build process will throw a bunch of warnings, but ultimately succeed:

    ```
    👀  ./packages/server/src/nonExistent.ts
    Module build failed (from ./node_modules/ts-loader/index.js):
    Error: ENOENT: no such file or directory, open '/Users/gregbrimble/projects/workers.sh/packages/server/src/nonExistent.ts'
    // ... and so on...
    ✨  Built successfully, built project size is 141 KiB.
    ```
    But the `dev` process itself will die shortly after:
    ```
    thread '<unnamed>' panicked at '--watch change message failed to send: "SendError(..)"', src/commands/build/wranglerjs/mod.rs:108:29
    ```

* Over the last couple of months, `cloudflared` has gone down a number of times, for various reasons. I know it's outwith your scope and control, and I might be misunderstanding things, but if Argo Tunnels don't work, `wrangler dev` becomes non-functional, and my development efforts come to a grinding halt, which is obviously frustrating. I don't know if there is much that can be done here, but I just wanted to express my concerns. I'm not sure if these issues have been isolated to `trycloudflare.com`'s free tunnels (and that premium tunnels are fine), or if the problems were across the board. But if the former, perhaps an option to allow users with a premium account to login and use their bandwidth would be beneficial? ("[We don’t guarantee any SLA or uptime of TryCloudflare](https://developers.cloudflare.com/argo-tunnel/trycloudflare/).") I don't know much about Argo Tunnels, and might have a complete misunderstanding of this (and/or be conflating an outage of `cloudflared` with the Workers preview service), so would be interested to hear your thoughts.

* I _think_ there might be a discrepency in behaviour between `wrangler dev` and production for `event.waitUntil()`. If I enqueue a Promise which takes a long time to resolve (but still much less than [the 30 seconds allowed on production](https://developers.cloudflare.com/workers/about/limits/)), `wrangler dev` will flakily fail with one of a number of different errors:

    ```
    Uncaught (in response)
    Error: internal error
      at async Promise.all (index 2)
      at async Promise.all (index 1)
      at async Object.t.api [as handler] (worker.js:46:75755)
      at async worker.js:46:31975
    ```

    ```
    A hanging Promise was canceled. This happens when the worker runtime is waiting for a Promise from JavaScript to resolve, but has detected that the Promise cannot possibly ever resolve because all code and events related to the Promise's request context have already finished.
    Uncaught (async)
    Error: Promise will never complete.
    ```

    ```
    Uncaught (in response)
    Error: internal error
      at async worker.js:46:31975
    ```

    And this usually results in a HTTP 503 with no body being returned in response to the request.

    But, if I publish the script to production, the response is seemingly always returned correctly, and the Promise, fulfilled.

    I don't have an MVP repoduction repo for this at the moment, so take the report with a pinch of salt, but can try throw one together if it's worth investigating.

Thank you all once again for `wrangler dev`. It is an excellent tool for developing with confidence. (Sorry for the dump of reports.)

**Comment by PierBover** [comment]
Can anyone share a tip to make webpack with `wrangler dev` faster?

I created a different webpack config during dev which disables minfication but as my project grows webpack is taking longer to build my project which sucks. Sometimes I need to wait up to 10 seconds for every change between the webpack build, the wasm download, and then the restart of the worker.

Usually with webpack projects I use the dev server which takes milliseconds to bundle new changes but when using `mode: 'development'` with wrangler I get this error which I don't know how to solve:

> Uncaught
> EvalError: Code generation from strings disallowed for this context

**Comment by EverlastingBugstopper** [comment]
Hey @PierBover - I have a PR in flight to fix the `wasm-pack` on every change that should go out once one of our dependencies updates. I know it's super annoying and I'll try to get a point release out with it soon, but like I said I'm waiting on a third party dep.

Your second idea of using `type = javascript` is what I would recommend, probably in conjunction with `webpack --watch`. We've sorta rolled our own JavaScript watching which does not have incremental compilation, which is why I have opened [this issue](https://github.com/cloudflare/wrangler/issues/1219).

As for disallowing `eval` with `devtool = none` this is actually the first I've heard of anybody running into this. Would you mind opening a new issue for this one? It might be worth including it in the default webpack configuration.

**Comment by PierBover** [comment]
Ok, I'll open that issue about `eval()`.

Thanks for your suggestions and your work on `dev` @EverlastingBugstopper !

**Comment by PierBover** [comment]
So after being a bit annoyed of having to manually restart `wrangler dev` randomly on every Rust panic I found a solution to restart wrangler automatically whenever it panics.

I'm using PM2 which is a process manager typically used in Node servers but can be used with any process.

https://pm2.keymetrics.io/

First install PM2 with:
```
sudo npm install pm2@latest -g
```
I installed it globally but I guess it can be installed in the project too.

Then to start a new PM2 process:
```
pm2 start /usr/local/bin/wrangler "wrangler" -- dev --env local --port 5555
```
* `pm2 start` starts a new process in the PM2 manager
* `/usr/local/bin/wrangler` the path to your wrangler binary
* `"wrangler"` the name of the new process. This can be anything you wish.
* `--` a separator to pass params to the binary
* `dev --env local --port 5555` my own params for `wrangler dev`.

After this PM2 will run in the background keeping wrangler alive no matter what.

To get the console log in realtime you need to `pm2 logs`.

To stop a process: `pm2 stop "wrangler"` (using the name of the process you used)

To kill PM2 and all its processes `pm2 kill`.

**Comment by EverlastingBugstopper** [comment]
@PierBover - nice workaround, I have a fix for this coming shortly in #1276 that should hopefully fix those annoying panics!

**Comment by PierBover** [comment]
Awesome @EverlastingBugstopper !

If you can alleviate the panics and the slow restarts my dev workflow with workers will already be better than using any of the other serverless solutions I've used.

**Comment by EverlastingBugstopper** [comment]
@GregBrimble - I'm sorry I haven't responded to your message! It's chocked full of a lot of really helpful information, I think you're probably the one with the most complicated setup.

Would you mind filing issues with steps to reproduce for each of the errors you've reported? As for cloudflared availability, it might help for you to use an alternative tool like [ngrok](https://ngrok.com/) as it may be more reliable

**Comment by ispivey** [comment]
@GregBrimble re: cloudflared disconnections, the team that manages cloudflared just released a new version (2020.5.1) that introduces a "quick reconnect mode" that is supposed to be much more resilient to disconnection events. You can see their changelog here: https://github.com/cloudflare/cloudflared/blob/master/RELEASE_NOTES

I'd suggest trying an upgrade of cloudflared, if you can!

**Comment by GregBrimble** [comment]
@ispivey, I'll check it out, thanks for the heads-up!

@EverlastingBugstopper, no worries—thanks for getting back to me! I'll find some minimal repro steps and get them filed later this week(/end) :)

**Comment by ispivey** [comment]
@GregBrimble @PierBover (and other watchers), check out the 1.9.1 release! @EverlastingBugstopper made a bunch of big performance and error handling improvements for `dev`. Curious to hear if this further improves your experience.

**Comment by PierBover** [comment]
Thanks @ispivey already using it!

So far it does improve the experience considerably.

**Comment by danbars** [comment]
Hey, `wrangler dev` seem to be very slow in picking up changes. I have to wait 30-60 seconds after saving a file before I see `Detected changes...` log line. 
This is consistent with every run.
I'm currently using 1.9.2, but it was the same in previous versions that I used (1.8.1 I think was the prev one)

**Comment by PierBover** [comment]
> I have to wait 30-60 seconds after saving a file before I see Detected changes... log line.

That's a lot @danbars . Personally I've never seen that. At the most it takes 2-3 seconds.

You should probably open a new issue with more details about your setup.

**Comment by peracto** [comment]
@EverlastingBugstopper,
I have an issue with wrangler dev ... When referencing the node 'crypto' package in a code module  - I get a 'panic'. If I comment out the 'require' and stub out the function - then all is well again.

If I publish my worker - then it performs as expected.

The odd thing is that the panic is raised without any of my code being execute. If I execute 'wrangler dev' - I get the usual messages - then after a few seconds, the panic is raised. At this point, I've not interacted with the worker.

Sometimes - if I mange to POST a message to the dev worker - then it does seem to work - and continues for a short while before randomly panic-ing. 
```
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Protocol("Connection reset without closing handshake")', src/commands/dev/socket.rs:46:27
stack backtrace:
   0:        0x106f2c00f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hdfed66b3e2e8ce60
   1:        0x106f533ee - core::fmt::write::h9de7b13cfaa80b03
   2:        0x106f22c97 - std::io::Write::write_fmt::h9197012b9d65a9a3
   3:        0x106f2e6da - std::panicking::default_hook::{{closure}}::h7b209cdc44c5777f
   4:        0x106f2e41c - std::panicking::default_hook::hf135bc502e6f77db
   5:        0x106f2ed48 - std::panicking::rust_panic_with_hook::h7967810bc33e523b
   6:        0x106f2e912 - rust_begin_unwind
   7:        0x106f72e3f - core::panicking::panic_fmt::h9eea9e4965bd189c
   8:        0x106f72d45 - core::option::expect_none_failed::ha68a01397bf24c9a
   9:        0x106831ed5 - <futures_util::stream::stream::for_each::ForEach<St,Fut,F> as core::future::future::Future>::poll::h80131d0f478b34b7
  10:        0x106930deb - <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll::hd70c8e3ef01a63e5
  11:        0x106867cf4 - <std::future::GenFuture<T> as core::future::future::Future>::poll::h6eebd6b4cfa53e9e
  12:        0x106942148 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h58d2bab017523f82
  13:        0x106893a0c - std::panicking::try::do_call::ha7942900b635f57b
  14:        0x106f3822b - __rust_maybe_catch_panic
  15:        0x106abec8d - tokio::runtime::task::harness::Harness<T,S>::poll::hdcda00bdd8bbf175
  16:        0x106ef960e - tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}::h314e68c2e244a164
  17:        0x106ef736e - std::thread::local::LocalKey<T>::with::h7fa5fdc8d103e6a4
  18:        0x106ef4cc4 - tokio::runtime::thread_pool::worker::Context::run_task::hd7662f38dfae470d
  19:        0x106ef4027 - tokio::runtime::thread_pool::worker::Context::run::h10d867d30c86b9ab
  20:        0x106f00c33 - tokio::macros::scoped_tls::ScopedKey<T>::set::hdac3c1ee6c79e370
  21:        0x106ef3e74 - tokio::runtime::thread_pool::worker::run::he028a4b011ba4e68
  22:        0x106efcb2d - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h0c298aee5190f512
  23:        0x106eed987 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hffadae7c67091bee
  24:        0x106ef23dc - std::panicking::try::do_call::h8c5d837d4f5baed6
  25:        0x106f3822b - __rust_maybe_catch_panic
  26:        0x106f04ed9 - tokio::runtime::task::harness::Harness<T,S>::poll::h1df49121de9e2268
  27:        0x106ee5066 - tokio::runtime::blocking::pool::Inner::run::h3414a9d5eb878407
  28:        0x106eeba62 - tokio::runtime::context::enter::h9144231ba50a7d70
  29:        0x106f01f9e - std::sys_common::backtrace::__rust_begin_short_backtrace::h3f0beb9c7f759ec9
  30:        0x106ef2330 - std::panicking::try::do_call::h49cfda0bcbc48286
  31:        0x106f3822b - __rust_maybe_catch_panic
  32:        0x106ee14f7 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h6d1f958516cfee5e
  33:        0x106f1923e - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h9a0848a2221632a4
  34:        0x106f377db - std::sys::unix::thread::Thread::new::thread_start::h9f834bdc6e39a9b2
  35:     0x7fff6f25b109 - _pthread_start
Error: panic
   0: backtrace::backtrace::trace
   1: backtrace::capture::Backtrace::new_unresolved
   2: failure::backtrace::internal::InternalBacktrace::new
   3: failure::backtrace::Backtrace::new
   4: <tokio::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll
   5: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
   6: <std::future::GenFuture<T> as core::future::future::Future>::poll
   7: std::thread::local::LocalKey<T>::with
   8: tokio::runtime::enter::Enter::block_on
   9: tokio::runtime::context::enter
  10: wrangler::commands::dev::gcs::dev
  11: wrangler::commands::dev::dev
  12: wrangler::run
  13: std::rt::lang_start::{{closure}}
  14: std::panicking::try::do_call
  15: __rust_maybe_catch_panic
  16: std::rt::lang_start_internal
  17: main
```

For reference this is the module that triggers the issue ...

```javascript
const crypto = require('crypto')
const _sign = (input, key) => crypto.createSign('RSA-SHA256').update(input).sign(key, 'base64')
const _escape = (str) => str.replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '')
const _base64 = (obj) => _escape(Buffer.from(obj).toString('base64'))

const header = _base64(JSON.stringify({alg: 'RS256'}))

module.exports = {
  sign(payload, key) {
    const input = header + '.' + _base64(JSON.stringify(payload))
    const buffer = Buffer.from(input)
    return input + '.' + _escape(_sign(buffer, key))
  }
}

```

I've been struggling to narrow down the issue, and have tried a few different versions of nodejs but to no avail. I'm hoping that you'll look at the panic backtrace, and spot the issue.

Any thoughts or insight ?

**Comment by EverlastingBugstopper** [comment]
Yeah that's correct. I tried to implement listening on both ipv4 + ipv6 a while back but couldn't get it to only listen on only loopback addresses - had to expose it over the local network which was a no-go from security perspective. You can specify ipv4 by running `wrangler dev --ip 127.0.0.1`.

**Comment by pyk** [comment]
It would be nice if `wrangler dev` support multiple webpack configurations.

The use case is when the user need to bundle static assets that runs in the browser.

For example:

```js
// Webpack config to bundle worker script
const webWorkerBundleConfig = {
    target: "webworker",
    // redacted
};

// Webpack config to bundle javascript assets that run in the browser
const javascriptBrowserBundleConfig = {
    target: "web",
    // redacted
};

module.exports = [javascriptBrowserBundleConfig, webWorkerBundleConfig];
```

Enabling this feature would greatly improve the developer experiences when developing Workers Sites locally.

**Comment by EverlastingBugstopper** [comment]
hey @pyk - that sounds like a more general wrangler feature request than something specific to `wrangler dev`. would you mind opening a separate issue for that?

**Comment by pyk** [comment]
@EverlastingBugstopper done #1397 👌

**Comment by WearyMonkeyFSF** [comment]
I'm trying to proxy to a local webpack server hosted, e.g.

```
self.addEventListener('fetch', event => {
  event.respondWith(fetch('http://localhost'));
});
```

And I'm being rick rolled:

![rick](https://user-images.githubusercontent.com/65217269/85502042-e3cd0500-b629-11ea-993e-72e00884e6bb.png)

Hilarious! But a bit of guidance on how to proxy to localhost for testing would be useful :)

**Comment by GregBrimble** [comment]
Hey @WearyMonkeyFSF,

Unfortunately, hitting localhost is not possible with wrangler just yet (see #901, #902 for progress). In the meantime, until the feature is added to wrangler, [I've thrown together a template repo](https://github.com/GregBrimble/cf-workers-typescript-template) which uses `cloudflared` to expose your localhost and then automatically connects wrangler to that.

It _should_ help bridge the gap until we add the feature to wrangler itself.

Let me know if you have any questions :)

**Comment by WearyMonkeyFSF** [comment]
Thanks @GregBrimble !

I slowly figured that out, but took me an hour or two.

Maybe you can put some instructions above the video? Just something to save future developers time.

Cheers!

**Comment by yet-another-dev** [comment]
I'm unable to get `wrangler dev` to return multiple `Set-Cookie` headers in a worker response. I've tried in both of the following ways: 

_explicitly appending them_
```js
// not working, only last get sent (`cookie2`)

resp.headers.append('Set-Cookie', 'cookie1=true')
resp.headers.append('Set-Cookie', 'cookie2=true')
```
_passing them in the options object of the response_
```js
// not working, only last get sent (`cookie2`)

headers: {
   'Set-Cookie': 'cookie1=true',
   'Set-Cookie': 'cookie2=true'
}
```

I've so far only been trying to get this to work using `wrangler dev`... perhaps it's a wider Workers issue/limitation? Perhaps I'm missing something obvious here?

**Comment by EverlastingBugstopper** [comment]
wow @yet-another-dev - _fantastic_ catch. thank you so much! can't believe this has not been caught before. i'll get a patch up shortly :)

**Comment by EverlastingBugstopper** [comment]
just released 1.10.3 that resolves this issue @yet-another-dev 😄 thanks again for the report

**Comment by yet-another-dev** [comment]
On the multiple cookie front (thanks for the super quick fix on this, @EverlastingBugstopper!), I've found that only the "explicitly appending" approach works (see https://github.com/cloudflare/wrangler/issues/1047#issuecomment-649705103). I assume this is intended (i.e. that the "passing them in the options object of the response" approach does not work by design)?

**Comment by ecerroni** [comment]
TLDR:
- deployed code works
- same code running wrangler dev does not work (no errors either)

---------
I am developing a Graphql server building upon https://github.com/signalnerve/workers-graphql-server

Using the base template with `wrangler dev` gave no issues. However as soon I started trying stitching local and remote schemas together I could not debug anymore locally.

If I do `wrangler publish` the changes are deployed and the production version on the edge behaves exactly as expected.

Locally though I keep getting redirected to example.com. No errors either although I guess there are errors swallowed by wrangler dev. There is something it does not like, but I cannot figure out what it is.

Any clues?

**Comment by EverlastingBugstopper** [comment]
hey @ecerroni - so what you're running into is probably due to the fact that `wrangler dev` is actually making requests via the Workers preview API (`wrangler dev` is not a "local" development server in that you need an internet connection for it to work). That means that anytime you want to make a request to something _else_ that's running on your local network, you'll need to expose that service to the public internet so the Cloudflare Workers runtime may make requests to it. 

@GregBrimble has an unofficial template for setting something this up [here](https://github.com/GregBrimble/cf-workers-typescript-template), but the gist of it is that it uses [cloudflared](https://developers.cloudflare.com/argo-tunnel/downloads) to create a public endpoint for the service running locally on your computer. It will then give you a public endpoint, and you can pass that hostname to `wrangler dev` via the `--host` flag. Hope this helps!

**Comment by ecerroni** [comment]
@EverlastingBugstopper Thank you. It seems straight forward. However, I tried it and it is not working for me. Maybe I am missing something obvious.

First I start the tunnel
```
$ cloudflared tunnel --url localhost:8787 --metrics localhost:8081
[INFO] Cannot determine default configuration path. No file [config.yml config.yaml] in [~/.cloudflared ~/.cloudflare-warp ~/cloudflare-warp /usr/local/etc/cloudflared /etc/cloudflared]
[INFO] Version 2020.6.5
[INFO] GOOS: linux, GOVersion: go1.12.9, GoArch: amd64
[INFO] Environment variables map[metrics:localhost:8081 proxy-dns-upstream:https://1.1.1.1/dns-query, https://1.0.0.1/dns-query url:localhost:8787]
[INFO] cloudflared will not automatically update when run from the shell. To enable auto-updates, run cloudflared as a service: https://developers.cloudflare.com/argo-tunnel/reference/service/
[INFO] Starting metrics server on 127.0.0.1:8081/metrics
[INFO] Proxying tunnel requests to http://localhost:8787
[INFO] Connected to VIE
[INFO] Each HA connection's tunnel IDs: map[0:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200]
[INFO] +-----------------------------------------------------------+
[INFO] |  Your free tunnel has started! Visit it:                  |
[INFO] |    https://offering-towers-satin-carol.trycloudflare.com  |
[INFO] +-----------------------------------------------------------+
[INFO] Route propagating, it may take up to 1 minute for your new route to become functional
[INFO] Connected to FRA
[INFO] Connected to VIE
[INFO] Each HA connection's tunnel IDs: map[0:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200 1:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200]
[INFO] +-----------------------------------------------------------+
[INFO] |  Your free tunnel has started! Visit it:                  |
[INFO] |    https://offering-towers-satin-carol.trycloudflare.com  |
[INFO] +-----------------------------------------------------------+
[INFO] Route propagating, it may take up to 1 minute for your new route to become functional
[INFO] Connected to FRA
[INFO] Each HA connection's tunnel IDs: map[0:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200 1:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200 2:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200]
[INFO] +-----------------------------------------------------------+
[INFO] |  Your free tunnel has started! Visit it:                  |
[INFO] |    https://offering-towers-satin-carol.trycloudflare.com  |
[INFO] +-----------------------------------------------------------+
[INFO] Route propagating, it may take up to 1 minute for your new route to become functional
[INFO] Each HA connection's tunnel IDs: map[0:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200 1:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200 2:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200 3:ecas8dbrglaz56yinus4ctdqiykxc8byti9hvzv0p3egc1ivy200]
[INFO] +-----------------------------------------------------------+
[INFO] |  Your free tunnel has started! Visit it:                  |
[INFO] |    https://offering-towers-satin-carol.trycloudflare.com  |
[INFO] +-----------------------------------------------------------+
[INFO] Route propagating, it may take up to 1 minute for your new route to become functional
```

Then I start wrangler dev with the public endpoint as the host
```
$ wrangler dev --host https://offering-towers-satin-carol.trycloudflare.com
`wrangler dev` is currently unstable and there are likely to be breaking changes!
For this reason, we cannot yet recommend using `wrangler dev` for integration testing.

Please submit any feedback here: https://github.com/cloudflare/wrangler/issues/1047
 Built successfully, built project size is 741 KiB.
 Listening on http://127.0.0.1:8787
 Detected changes...
 Ignoring stale first change
[INFO] Remote schema ready
[2020-06-30 11:31:43] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:43] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:43] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:42] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:42] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:41] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:41] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:41] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:40] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:40] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:40] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:40] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:40] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:39] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:39] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:39] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:39] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:46] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:46] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:45] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:45] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:45] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:45] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:44] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:44] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:44] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:44] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:44] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:43] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:43] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:43] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:42] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:42] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:42] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
[2020-06-30 11:31:42] GET offering-towers-satin-carol.trycloudflare.com/___graphql HTTP/1.1 403 Forbidden
thread 'tokio-runtime-worker' panicked at 'Could not determine status code of response', src/commands/dev/gcs/headers.rs:73:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'tokio-runtime-worker' panicked at 'Could not determine status code of response', src/commands/dev/gcs/headers.rs:73:18
thread 'tokio-runtime-worker' panicked at 'Could not determine status code of response', src/commands/dev/gcs/headers.rs:73:18
thread 'tokio-runtime-worker' panicked at 'Could not determine status code of response', src/commands/dev/gcs/headers.rs:73:18
```

This is the screen I see when I try to access `http://127.0.0.1/___graphql` or any other url on localhost.
![url](https://user-images.githubusercontent.com/11174327/86110514-efa34600-bac5-11ea-9dbe-37e6b0de1df2.png)

**Comment by Autarc** [comment]
Going through the thread I couldn't see if it was already mentioned before but while the `dev` mode is great for local development there is a small issue:

```js
request.header.get('Accept-Encoding'); 
```

The output in the `wrangler dev` runtime returns `gzip, deflate, br` while after deployment it will always be `gzip` [as its set internally](https://support.cloudflare.com/hc/en-us/articles/200168086-Does-Cloudflare-compress-resources-). While it is certainly a specific use case it might be good to have the behavior aligned. Even though it would be great to be able to access the information somehow it currently doesn't seem to be possible so having it statically set during development help to be aware of the it in time. 

On another note it seems that there will be [a `script.js` file](https://github.com/cloudflare/wrangler/blob/master/src/wranglerjs/bundle.rs#L63) created in the `worker` directory. Since the webpack configuration already has an output path perhaps that could be used for configuring the destination as it was a bit surprising to see it generated in that place 😅

**Comment by EverlastingBugstopper** [comment]
@ecerroni 

> cloudflared tunnel --url localhost:8787 --metrics localhost:8081

this won't work because you're exposing `wrangler dev` to the public internet, and then trying to use it as a back end - i assumed you wanted to make requests to some other sort of back end you were developing side-by-side with your worker?

perhaps i'm misunderstanding what you're trying to do - could you open a new issue and list out steps to reproduce the issue you're having (especially helpful if you can link a repository i can check out and futz around with)

---

@Autarc - could you file a new issue for the request header thing you're seeing and we'll take a look?

as for the `script.js` in the `worker` dir is tracked in #1046

**Comment by Autarc** [comment]
@EverlastingBugstopper 

Sure created it here: https://github.com/cloudflare/wrangler/issues/1425

Also thanks for clarifying the `script.js` file and referring to the existing issue. Keep up the great work!

**Comment by ecerroni** [comment]
@EverlastingBugstopper 
> 
> > cloudflared tunnel --url localhost:8787 --metrics localhost:8081
> 
> this won't work because you're exposing `wrangler dev` to the public internet, and then trying to use it as a back end - i assumed you wanted to make requests to some other sort of back end you were developing side-by-side with your worker?
> 
> perhaps i'm misunderstanding what you're trying to do - could you open a new issue and list out steps to reproduce the issue you're having (especially helpful if you can link a repository i can check out and futz around with)

Trying recreating the issue from scratch using a different remote schema I realized that the culprit is in the json file I am trying to require and pass in the remoteExecutableSchema.

If the file is small there is no problem.
If the size is around 100kB it does not work on wrangler dev, but it does after publishing (although I might retry a couple of times).
If the file is larger then wrangler dev is not working and I cannot publish anymore. I always get script has timed out.

At this point I am confused and I do not know if it is either me doing something wrong in the code or it is the graphql's makeRemoteExecutableSchema having issues with larger files or I cannot load large local json files (under `.src/`) with workers or maybe I need additional configuration in webpack to handle this case :man_shrugging:

**Comment by EverlastingBugstopper** [comment]
🤔 - hmmmm.... unfortunately, i have no clue what your issue is from reading this. if you could link a repository in a new issue with detailed steps to reproduce, i'll be more than happy to take a look. for now, logging off for the long weekend here in the US!

**Comment by ecerroni** [comment]
@EverlastingBugstopper 

Sure, enjoy your weekend :)

In the meantime, I created an issue [here](https://github.com/cloudflare/wrangler/issues/1435) with a reproducing repo

**Comment by normanr** [comment]
> The following command doesn't work:
> 
> ```shell
> NODE_OPTIONS="--inspect" wrangler dev
> ```

There are at least two node instances, so I found that making the inspector use a random port with `NODE_OPTIONS="--inspect=:0"` at least gets things running successfully. I haven't been able to inspect my code though. I suspect because as EverlastingBugstopper said "Cloudflare Workers are not executed within the context of a Node runtime". So the node instances you can attach to are running other things, not your workers.

**Comment by EverlastingBugstopper** [comment]
Hey @normanr - we have an issue for that here: #946, but it has not been prioritized and i'm not sure when it will be

**Comment by normanr** [comment]
ack, I've subscribed to it, thanks!

**Comment by mskd12** [comment]
@EverlastingBugstopper is it possible to automatically set some environment variables only when using `wrangler dev`? Sounds like something basic, so sorry if I missed it in the docs somehow.

**Comment by PierBover** [comment]
@mskd12 you can set up different environments and create vars for each environment.

See the docs:

https://developers.cloudflare.com/workers/tooling/wrangler/environments/

**Comment by mskd12** [comment]
@PierBover I did check it, but I was not sure about the interaction between environments and `wrangler dev`. Does calling `wrangler dev` automatically set the environment to `dev`?

**Comment by EverlastingBugstopper** [comment]
@mskd12 - no, environments work the same exact way they do for any other command. if you dont specify an `--env`, it will use your top level configuration. if you want to use the `dev` environment you must `wrangler dev --env dev`

**Comment by SupremeTechnopriest** [comment]
Would be nice to be able to define a port in `workers dev --host https://localhost:3000`.  Currently the port is stripped out.

**Comment by normanr** [comment]
I think you need to pass the port using the --port flag (#1477 made it so that it can also be set in wrangler.toml).

**Comment by EverlastingBugstopper** [comment]
@SupremeTechnopriest there are two things going on here

1) `--host` is meant for setting your upstream host. as in, when you access `request.url` in your Worker, what host is it? You can't really set `--host` to anything that isn't running on the public Internet so if you want `request.url` to go to something running on port 3000 you must first expose it to the Internet with a tool like cloudflared or ngrok.

2) If what you really want to do is change the port that `wrangler dev` is running on, then you should use the `--port` flag.

**Comment by SupremeTechnopriest** [comment]
@EverlastingBugstopper So even though I'm running `dev` the worker is still running in the cloudflare backplane and it is just forwarding the logs to my machine? I assumed the worker was actually running on my machine, therefore localhost as a upstream host would be ok.  I ended up just using a tunnel.

**Comment by EverlastingBugstopper** [comment]
Yup, running on Cloudflare's servers

**Comment by SupremeTechnopriest** [comment]
@EverlastingBugstopper Thanks for the info!  Makes sense now.
